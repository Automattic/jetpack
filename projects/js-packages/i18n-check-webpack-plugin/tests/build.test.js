const fs = require( 'fs' );
const path = require( 'path' );
const webpack = require( 'webpack' );

const fixturesPath = path.join( __dirname, 'fixtures' );
const configFixtures = fs.readdirSync( fixturesPath ).sort();

/**
 * Find all files in a path.
 *
 * @param {string} dir - Path.
 * @returns {string[]} Promise resolving to a list of files.
 */
function lsFiles( dir ) {
	const ret = [];
	for ( const dirent of fs.readdirSync( dir, { withFileTypes: true } ) ) {
		if ( dirent.isDirectory() ) {
			for ( const file of lsFiles( path.join( dir, dirent.name ) ) ) {
				ret.push( path.join( dirent.name, file ) );
			}
		} else {
			ret.push( dirent.name );
		}
	}
	return ret;
}

test.each( configFixtures )(
	'Webpack `%s`',
	async fixture => {
		const testdir = path.join( fixturesPath, fixture );
		const builddir = path.join( testdir, 'dist' );
		fs.rmSync( builddir, { force: true, recursive: true } );

		const config = require( path.join( testdir, 'webpack.config.js' ) );
		if ( Array.isArray( config ) ) {
			for ( const c of config ) {
				c.context = c.context || testdir;
				c.output.path = builddir;
			}
		} else {
			config.context = config.context || testdir;
			config.output.path = builddir;
		}

		const [ err, stats ] = await new Promise( resolve => {
			// eslint-disable-next-line no-shadow
			webpack( config, ( err, stats ) => {
				resolve( [ err, stats ] );
			} );
		} );

		expect( err ).toMatchSnapshot( 'Webpack build error' );

		// Webpack's error messages may contain full paths. Munge those.
		const mungePaths = x => {
			if ( typeof x === 'string' ) {
				return x.replace( __dirname, '/path/to/i18n-check-webpack-plugin/tests' );
			}
			if ( Array.isArray( x ) ) {
				return x.map( mungePaths );
			}
			if ( typeof x === 'object' && Object.getPrototypeOf( x ) === Object.prototype ) {
				return Object.fromEntries(
					Object.entries( x ).map( ( [ k, v ] ) => [ k, mungePaths( v ) ] )
				);
			}
			return x;
		};
		const statsOut = stats
			? mungePaths( stats.toJson( { all: false, errors: true, warnings: true } ) )
			: null;
		expect( statsOut ).toMatchSnapshot( 'Webpack build stats' );

		expect( lsFiles( builddir ) ).toMatchSnapshot( 'Webpack build files' );

		if ( fixture === 'dropped-is-ok' ) {
			const contents = fs.readFileSync( path.join( builddir, 'main.js' ), { encoding: 'utf8' } );
			// eslint-disable-next-line jest/no-conditional-expect
			expect( contents ).toEqual( expect.stringContaining( 'This is production' ) );
			// eslint-disable-next-line jest/no-conditional-expect
			expect( contents ).toEqual( expect.not.stringContaining( 'This is not production' ) );
		}
	},
	15000
);
