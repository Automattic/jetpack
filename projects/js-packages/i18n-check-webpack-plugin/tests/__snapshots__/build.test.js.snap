// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webpack \`basic-test\`: Webpack build error 1`] = `null`;

exports[`Webpack \`basic-test\`: Webpack build files 1`] = `
[
  "main.asset.php",
  "main.js",
]
`;

exports[`Webpack \`basic-test\`: Webpack build stats 1`] = `
{
  "errors": [],
  "warnings": [],
}
`;

exports[`Webpack \`complex-test\`: Webpack build error 1`] = `null`;

exports[`Webpack \`complex-test\`: Webpack build files 1`] = `
[
  "async.js",
  "css.asset.php",
  "css.css",
  "css.js",
  "css.rtl.css",
  "main.asset.php",
  "main.js",
]
`;

exports[`Webpack \`complex-test\`: Webpack build stats 1`] = `
{
  "errors": [],
  "warnings": [],
}
`;

exports[`Webpack \`concatenated-modules\`: Webpack build error 1`] = `null`;

exports[`Webpack \`concatenated-modules\`: Webpack build files 1`] = `
[
  "main.asset.php",
  "main.js",
]
`;

exports[`Webpack \`concatenated-modules\`: Webpack build stats 1`] = `
{
  "errors": [
    {
      "message": "main.js: Translator comments have gone missing for "A"
 - Translators: A",
    },
    {
      "message": "main.js: Translator comments have gone missing for "B"
 - Translators: B",
    },
  ],
  "warnings": [],
}
`;

exports[`Webpack \`known-problems\`: Webpack build error 1`] = `null`;

exports[`Webpack \`known-problems\`: Webpack build files 1`] = `
[
  "main.asset.php",
  "main.js",
]
`;

exports[`Webpack \`known-problems\`: Webpack build stats 1`] = `
{
  "errors": [
    {
      "message": "main.js:2:92: msgid argument is not a string literal: __(o?"Foo is set":"Foo is not set","domain")",
    },
    {
      "message": "main.js: Translator comments have gone missing for "The comment will be lost when the assignment is merged into the above const" (context "context")
 - Translators: This comment will be lost",
    },
    {
      "message": "main.js: Optimization seems to have broken the following translation strings:
 - "Foo is not set"
 - "Foo is set"",
    },
  ],
  "warnings": [],
}
`;

exports[`Webpack \`multiple-domains\`: Webpack build error 1`] = `null`;

exports[`Webpack \`multiple-domains\`: Webpack build files 1`] = `
[
  "main.asset.php",
  "main.js",
]
`;

exports[`Webpack \`multiple-domains\`: Webpack build stats 1`] = `
{
  "errors": [
    {
      "message": "main.js: Multiple textdomains are used: "domain1", "domain2"
You may want to use @automattic/babel-plugin-replace-textdomain to fix that.",
    },
  ],
  "warnings": [],
}
`;

exports[`Webpack \`options-filter\`: Webpack build error 1`] = `null`;

exports[`Webpack \`options-filter\`: Webpack build files 1`] = `
[
  "array/main.asset.php",
  "array/main.js",
  "function/main.asset.php",
  "function/main.js",
  "regex/main.asset.php",
  "regex/main.js",
  "string/main.asset.php",
  "string/main.js",
  "undefined/main.asset.php",
  "undefined/main.js",
]
`;

exports[`Webpack \`options-filter\`: Webpack build stats 1`] = `
{
  "children": [
    {
      "errors": [
        {
          "message": "string/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
        },
        {
          "message": "string/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
        },
        {
          "message": "string/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
        },
        {
          "message": "string/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
        },
        {
          "message": "string/main.js: Optimization seems to have broken the following translation strings:
 - "string is not set"
 - "string is set"",
        },
      ],
      "name": "string",
      "warnings": [],
    },
    {
      "errors": [
        {
          "message": "regex/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
        },
        {
          "message": "regex/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
        },
        {
          "message": "regex/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
        },
        {
          "message": "regex/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
        },
        {
          "message": "regex/main.js: Optimization seems to have broken the following translation strings:
 - "regex is not set"
 - "regex is set"",
        },
      ],
      "name": "regex",
      "warnings": [],
    },
    {
      "errors": [
        {
          "message": "function/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
        },
        {
          "message": "function/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
        },
        {
          "message": "function/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
        },
        {
          "message": "function/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
        },
        {
          "message": "function/main.js: Optimization seems to have broken the following translation strings:
 - "func is not set"
 - "func is set"",
        },
      ],
      "name": "function",
      "warnings": [],
    },
    {
      "errors": [
        {
          "message": "array/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
        },
        {
          "message": "array/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
        },
        {
          "message": "array/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
        },
        {
          "message": "array/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
        },
        {
          "message": "array/main.js: Optimization seems to have broken the following translation strings:
 - "arr is not set"
 - "arr is set"",
        },
      ],
      "name": "array",
      "warnings": [],
    },
    {
      "errors": [
        {
          "message": "undefined/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
        },
        {
          "message": "undefined/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
        },
        {
          "message": "undefined/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
        },
        {
          "message": "undefined/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
        },
        {
          "message": "undefined/main.js: Optimization seems to have broken the following translation strings:
 - "arr is not set"
 - "arr is set"
 - "func is not set"
 - "func is set"
 - "regex is not set"
 - "regex is set"
 - "string is not set"
 - "string is set"",
        },
      ],
      "name": "undefined",
      "warnings": [],
    },
  ],
  "errors": [
    {
      "compilerPath": "string",
      "message": "string/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
    },
    {
      "compilerPath": "string",
      "message": "string/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
    },
    {
      "compilerPath": "string",
      "message": "string/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
    },
    {
      "compilerPath": "string",
      "message": "string/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
    },
    {
      "compilerPath": "string",
      "message": "string/main.js: Optimization seems to have broken the following translation strings:
 - "string is not set"
 - "string is set"",
    },
    {
      "compilerPath": "regex",
      "message": "regex/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
    },
    {
      "compilerPath": "regex",
      "message": "regex/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
    },
    {
      "compilerPath": "regex",
      "message": "regex/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
    },
    {
      "compilerPath": "regex",
      "message": "regex/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
    },
    {
      "compilerPath": "regex",
      "message": "regex/main.js: Optimization seems to have broken the following translation strings:
 - "regex is not set"
 - "regex is set"",
    },
    {
      "compilerPath": "function",
      "message": "function/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
    },
    {
      "compilerPath": "function",
      "message": "function/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
    },
    {
      "compilerPath": "function",
      "message": "function/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
    },
    {
      "compilerPath": "function",
      "message": "function/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
    },
    {
      "compilerPath": "function",
      "message": "function/main.js: Optimization seems to have broken the following translation strings:
 - "func is not set"
 - "func is set"",
    },
    {
      "compilerPath": "array",
      "message": "array/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
    },
    {
      "compilerPath": "array",
      "message": "array/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
    },
    {
      "compilerPath": "array",
      "message": "array/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
    },
    {
      "compilerPath": "array",
      "message": "array/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
    },
    {
      "compilerPath": "array",
      "message": "array/main.js: Optimization seems to have broken the following translation strings:
 - "arr is not set"
 - "arr is set"",
    },
    {
      "compilerPath": "undefined",
      "message": "undefined/main.js:1:77: msgid argument is not a string literal: __(t?"arr is set":"arr is not set","domain")",
    },
    {
      "compilerPath": "undefined",
      "message": "undefined/main.js:1:187: msgid argument is not a string literal: __(t?"func is set":"func is not set","domain")",
    },
    {
      "compilerPath": "undefined",
      "message": "undefined/main.js:1:376: msgid argument is not a string literal: __(t?"regex is set":"regex is not set","domain")",
    },
    {
      "compilerPath": "undefined",
      "message": "undefined/main.js:1:490: msgid argument is not a string literal: __(t?"string is set":"string is not set","domain")",
    },
    {
      "compilerPath": "undefined",
      "message": "undefined/main.js: Optimization seems to have broken the following translation strings:
 - "arr is not set"
 - "arr is set"
 - "func is not set"
 - "func is set"
 - "regex is not set"
 - "regex is set"
 - "string is not set"
 - "string is set"",
    },
  ],
  "warnings": [],
}
`;

exports[`Webpack \`options-warnOnly\`: Webpack build error 1`] = `null`;

exports[`Webpack \`options-warnOnly\`: Webpack build files 1`] = `
[
  "main.asset.php",
  "main.js",
]
`;

exports[`Webpack \`options-warnOnly\`: Webpack build stats 1`] = `
{
  "errors": [],
  "warnings": [
    {
      "message": "main.js:1:77: msgid argument is not a string literal: __(t?"X is set":"X is not set","domain")",
    },
    {
      "message": "main.js: Optimization seems to have broken the following translation strings:
 - "X is not set"
 - "X is set"",
    },
  ],
}
`;

exports[`Webpack \`pruned-branches\`: Webpack build error 1`] = `null`;

exports[`Webpack \`pruned-branches\`: Webpack build files 1`] = `
[
  "main.asset.php",
  "main.js",
]
`;

exports[`Webpack \`pruned-branches\`: Webpack build stats 1`] = `
{
  "errors": [
    {
      "message": "main.js: Optimization seems to have broken the following translation strings:
 - "another thing"",
    },
  ],
  "warnings": [],
}
`;
